### ISSUE TITLE
Apply custom notification special treatment not only to payload but other fields in httpCustom

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#2473]
Currently, Orion applies a decoding step to httpCustom.payload in order to render forbidden chars in custom notification payloads, as described at https://fiware-orion.readthedocs.io/en/develop/user/forbidden_characters/index.html#custom-payload-special-treatment

However, maybe this should be also applied to others subfields in httpCustom. In particular, we have found cases from users that would need to apply it to httpCustom.headers map value.

For example, the following have a problematic header value, as `;` and `=` are forbidden.

```
      "headers": {
            "Content-Type": "application/json; charset=utf-8",
            "Fiware-Service": "foo",
            "Fiware-ServicePath": "/bar"
        }
```

Considering the above approach, this can be rewritten as:

```
      "headers": {
            "Content-Type": "application/json%3B charset%3Dutf-8",
            "Fiware-Service": "foo",
            "Fiware-ServicePath": "/bar"
        }
```

**Comment by fgalan** [comment]
Another important case is "Authorization" header, which uses the `=` sign (see https://github.com/telefonicaid/fiware-orion/issues/2894).

**Comment by fgalan** [comment]
Looking to the code, it seems to be a matter of modifying `buildSenderParamsCustom()` to apply the unscape to other fields different from payload. This is the way in which payload is processed:

```
      macroSubstitute(&payload, httpInfo.payload, ce);
      char* pload  = curl_unescape(payload.c_str(), payload.length());
      payload      = std::string(pload);
      ...
      free(pload);
```

The importan statement is `curl_unescape()`.

For instance, to apply to a header value:

```
      std::string key   = it->first;
      std::string value = it->second;

      macroSubstitute(&key,   it->first, ce);
      macroSubstitute(&value, it->second, ce);

      if (key == "")
      {
        // To avoid empty header name
        continue;
      }

      headers[key] = value;
```

would be something like that:

```      
      std::string key   = it->first;
      std::string value = it->second;

      macroSubstitute(&key,   it->first, ce);
      macroSubstitute(&value, it->second, ce);

      char* vlue  = curl_unescape(value.c_str(), value.length()); // <---- NEW

      if (key == "")
      {
        // To avoid empty header name
        continue;
      }

      headers[key] = std::string(vlue); // <---- MODIFIED
      free(vlue);    // <----- NEW
```

**Comment by cdupont** [comment]
@fgalan to be honest, this encoding looks a bit weird... I think it's better to apply URL encoding to all the field (but it's difficult to read/write), or we don't apply it at all...
Anyway Is it possible to establish an exception for httpCustom fields? Is it a security issue?

**Comment by fgalan** [comment]
Yes, forbidden characters are to avoid script injections attack in some circustances (e.g. cross domain to co-located web servers in the same hot that CB). Thus, don't apply it at all could be risky.

**Comment by fgalan** [comment]
This issue is not especifically related with NGISv2, so APIv2 label is removed. The issue keeps opened.

CC: @jmcanterafonseca

**Comment by fgalan** [comment]
Similar issue https://github.com/telefonicaid/fiware-orion/issues/4071

**Comment by fgalan** [comment]
PR #4072 included in Orion 3.6.0 (associated to issue #4071) solves this for `headers`.

With regards to `url` and `qs` I understand it doesn't make sense, as they are used to compose the URL to send the notification, where having URL encoded chars is what the user wants (so decoding them could be counterproductive).

Thus, this issue is going to be closed.
